docs/proposals/20230407-managed-k8s-capi-contract-changes.md

```diff
@@ -0,0 +1,380 @@
+
+These broadly follow the existing separation within CAPI.
+
+However, for many Managed Kubernetes services this will require less than ideal code in the controllers to retrieve the control plane endpoint from the `<Infra>ControlPlane` kind and report it back via the ControlPlaneEndpoint property on the `<Infra>Cluster` to satisfy CAPI contracts.
+
+To give an idea what this means:
+- `<Infra>Cluster` watches the control plane and vice versa
+- `<Infra>Cluster` controller create base infra and sets Ready = true
+- `<Infra>ControlPlane` waits for `<Infra>Cluster` to be Ready
+- `<Infra>ControlPlane` creates an instance of the managed k8s service
+- `<Infra>ControlPlane` gets the API server endpoint from the managed k8s service and stores it in the CRD instance
+- `<Infra>Cluster` is watching for changes to `<Infra>ControlPlane` and if the "api server endpoint" on the `<Infra>ControlPlane` CRD instance is not empty then:
+  - Map `<Infra>ControlPlane` to `<Infra>Cluster` and queue event
+  - `<Infra>Cluster` reconciler loop gets the `<Infra>ControlPlane` CRD instance and takes the value for "api server endpoint" and populates `ControlPlaneEndpoint` on the `<Infra>Cluster` CRD instance.
+  - (which will then cause the reconciler for `<Infra>ControlPlane` to run... again)
+
+The implementation of the controllers for Managed Kubernetes would be simplified if there was an option to report the ControlPlaneEndpoint via `<Infra>ControlPlane` instead. Below we will outline two new flows that reduce much of the complexity of the above, while allowing Managed Kubernetes providers to represent their services intuitively.
+
+### Two New Flows
+
+#### Flow 1: `<Infra>Cluster` and `<Infra>ControlPlane`, with `ControlPlaneEndpoint` reported via `<Infra>ControlPlane`
+
+We will describe a CRD composition that adheres to the original separation of concerns of the different provider types as documented in the Cluster API documentation, with a different API Server endpoint reporting flow.
+
+As described above, at present the control plane endpoint must be returned via the `ControlPlaneEndpoint` field on the spec of the `<Infra>Cluster` [reference here](https://cluster-api.sigs.k8s.io/developer/providers/cluster-infrastructure.html). This is OK for self-managed clusters, as a load balancer is usually created as part of the reconciliation. But with Managed Kubernetes services the API Server endpoint usually comes from the service directly, which means that the `<Infra>Cluster` has to get the `ControlPlaneEndpoint` from the managed service so that it can be reported back to CAPI. In practice, this results in `<Infra>Cluster` watching the `<Infra>ControlPlane` and the `<Infra>ControlPlane` watching the `<Infra>Cluster`, and without care this can cause event storms in the CAPI management cluster.
```

Comment:
Continuing from the point above, it might be the case here that we didn't fully separate concerns properly so far in terms of endpoints for a Cluster. At some point there was also some interest in expressing load balancers in separate entities.

Instead of supporting endpoints through the control plane, have we considered the option to change the contract between the infrastructure providers and Cluster API controllers to have ways to gather endpoint information through a new well-defined CRD instead?

At a high level:
- Define something like `ClusterEndpoint` CRD, which should _eventually_ be created by another controller.
- Infrastructure provider CRDs should have controllers cooperate effectively internally depending if the endpoints are coming from the InfraCluster or InfraControlPlane resource.
- Eventually, we could support having multiple endpoints per cluster, for example to satisfy the private load balancer talks we've had in the past.

Reply:
Thanks for expanding! (😀 see my comment above)

This sounds really interesting. We didn't really expand the scope to include new CRDs. At the beginning when considering creating a native `ManagedCluster` CRD (which would obviously be a new CRD) to represent managed k8s natively in CAPI the community was convinced that there was a path forward (credit to @fabriziopandini) that was 100% back-compat but provided sufficent flexibility for current managed k8s use cases to remove some of the difficulty that implementers have encountered using existing CAPI spec and behaviors.

This idea would also perhaps accommodate multi-cluster scenarios.

Seems cool but out of scope or this proposal. Do we have an issue tracking this and the use-cases it might solve (or make easier)?

Reply:
> Seems cool but out of scope or this proposal

Not sure if totally out of scope, it's really a different contract than we have today, and it can developed to be fully backward compatible. As in, we can keep the current field in place until next API revision, and automatically create the new resource for InfraCluster that are not yet adhering to the contract.

Reply:
It's not strictly out of scope for the problem, you're correct, but during the process of creating this proposal consensus formed around preferring these behavioral changes over the addition of new CRDs.